src/main/java/uk/gov/digital/ho/hocs/queue/common/MigrationQueueListener.java

```diff
@@ -1,36 +1,55 @@
 package uk.gov.digital.ho.hocs.queue.common;
 
 import lombok.extern.slf4j.Slf4j;
-import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.cloud.aws.messaging.listener.SqsMessageDeletionPolicy;
 import org.springframework.cloud.aws.messaging.listener.annotation.SqsListener;
+import org.springframework.context.annotation.Profile;
 import org.springframework.messaging.handler.annotation.Header;
 import org.springframework.stereotype.Service;
+import uk.gov.digital.ho.hocs.service.MessageLogService;
 
 import java.util.List;
+import java.util.UUID;
 
 @Service
 @Slf4j
-@ConditionalOnProperty(name = "case-creator.mode", havingValue = "migration", matchIfMissing = false)
+@Profile("migration")
 public class MigrationQueueListener {
 
-    private final List<BaseMessageHandler> queueMessageHandlers;
+    private final List<MessageHandler> queueMessageHandlers;
 
-    public MigrationQueueListener(List<BaseMessageHandler> queueMessageHandlers) {
+    private final MessageLogService messageLogService;
+
+    private final boolean shouldIgnoreMessages;
+
+    public MigrationQueueListener(List<MessageHandler> queueMessageHandlers,
+                                  MessageLogService messageLogService,
+                                  @Value("${aws.sqs.ignore-messages:false}") boolean shouldIgnoreMessages) {
         this.queueMessageHandlers = queueMessageHandlers;
+        this.messageLogService = messageLogService;
+        this.shouldIgnoreMessages = shouldIgnoreMessages;
     }
 
     @SqsListener(value = "${aws.sqs.case-migrator.url}", deletionPolicy = SqsMessageDeletionPolicy.NO_REDRIVE)
-    public void onMigrationEvent(String message, @Header("MessageId") String messageId) throws Exception {
-        for (BaseMessageHandler messageHandler :
-                queueMessageHandlers) {
-            if (messageHandler.getMessageType().equals(MessageTypes.MIGRATION)) {
-                if (!messageHandler.shouldIgnoreMessage()) {
-                    messageHandler.handleMessage(message, messageId);
-                } else {
-                    log.warn("Message flagged to ignore: {}", messageId);
+    public void onMigrationEvent(String message,
+                                 @Header("MessageId") String messageId,
+                                 @Header(value = "ExternalReference", required = false) UUID externalReference) throws Exception {
+        // Create message log entry
+        messageLogService.createMessageLogEntry(messageId, externalReference, message);
+
+        if (shouldIgnoreMessages) {
+            log.warn("Message flagged to ignore: {}", messageId);
+        } else {
+            for (MessageHandler messageHandler :
+                    queueMessageHandlers) {
+                if (messageHandler.getMessageType().equals(MessageTypes.MIGRATION)) {
```

Comment:
I'd like to maybe pass this through a header too for verification. As we should be only spinning up one MessageHandler per deployment the `foreach` loop could likely be removed too.

Reply:
Agree we don't need the for loop here and this could be simplified to throw an exception if the message shouldn't be on this queue?

Reply:
This has been rectified here - there are fewer improvements but OOS for now.